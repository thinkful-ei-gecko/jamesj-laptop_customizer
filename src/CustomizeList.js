import React from 'react'
import slugify from 'slugify'
import CustomizeItem from './CustomizeItem'

export default function CustomizeList(props) {
  return Object.keys(props.features).map((feature, idx) => {
    const featureHash = feature + '-' + idx
    const options = props.features[feature].map(item => {
      const itemHash = slugify(JSON.stringify(item))
      console.log('CustomizeList props are:', props)
      console.log('the item hash:', itemHash)
      return (
        <CustomizeItem
          key={itemHash}
          itemHash={itemHash}
          feature={feature}
          {...item}
          selected={props.selected}
          updateFeature={props.updateFeature}
        />
      )
    })

    return (
      <fieldset className="feature" key={featureHash}>
        <legend className="feature__name">
          <h3>{feature}</h3>
        </legend>
        {options}
      </fieldset>
    )
  })
}
