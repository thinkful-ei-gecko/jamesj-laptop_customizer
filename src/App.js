import React, { Component } from 'react'

// Normalizes string as a slug - a string that is safe to use
// in both URLs and html attributes

import './App.css'
import CustomizeList from './CustomizeList'
import CartList from './CartList'

// This object will allow us to
// easily convert numbers into US dollar values
const USCurrencyFormat = new Intl.NumberFormat('en-US', {
  style: 'currency',
  currency: 'USD',
})

class App extends Component {
  constructor() {
    super()
    this.state = {
      features: {
        Processor: [
          {
            name: '17th Generation Intel Core HB (7 Core with donut spare)',
            cost: 700,
          },
          {
            name: 'Professor X AMD Fire Breather with sidewinder technology',
            cost: 1200,
          },
        ],
        'Operating System': [
          {
            name: 'Ubuntu Linux 16.04',
            cost: 200,
          },
          {
            name: 'Bodhi Linux',
            cost: 300,
          },
        ],
        'Video Card': [
          {
            name: 'Toyota Corolla 1.5v',
            cost: 1150.98,
          },
          {
            name: 'Mind mild breeze 2000',
            cost: 1345,
          },
        ],
        Display: [
          {
            name: '15.6" UHD (3840 x 2160) 60Hz Bright Lights and Knobs',
            cost: 1500,
          },
          {
            name: '15.3" HGTV (3840 x 2160) Home makeover edition',
            cost: 1400,
          },
        ],
      },
      selected: {
        Processor: {
          name: '17th Generation Intel Core HB (7 Core with donut spare)',
          cost: 700,
        },
        'Operating System': {
          name: 'Ubuntu Linux 16.04',
          cost: 200,
        },
        'Video Card': {
          name: 'Toyota Corolla 1.5v',
          cost: 1150.98,
        },
        Display: {
          name: '15.6" UHD (3840 x 2160) 60Hz Bright Lights and Knobs',
          cost: 1500,
        },
      },
    }
  }

  updateFeature = (feature, newValue) => {
    const selected = Object.assign({}, this.state.selected)
    selected[feature] = newValue
    this.setState({
      selected,
    })
  }

  render() {
    return (
      <div className="App">
        <header>
          <h1>ELF Computing | Laptops</h1>
        </header>
        <main>
          <form className="main__form">
            <h2>Customize your laptop</h2>
            <CustomizeList
              features={this.state.features}
              formatUS={USCurrencyFormat}
              selected={this.state.selected}
              updateFeature={(feature, newValue) =>
                this.updateFeature(feature, newValue)
              }
            />
          </form>
          <section className="main__summary">
            <h2>Your cart</h2>
            <CartList
              formatUS={USCurrencyFormat}
              selected={this.state.selected}
            />
          </section>
        </main>
      </div>
    )
  }
}

export default App
