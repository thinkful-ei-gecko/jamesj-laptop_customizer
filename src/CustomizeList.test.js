import React from 'react'
import ReactDOM from 'react-dom'
import CustomizeList from './CustomizeList'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { USCurrencyFormat, updateFeature } from './App'

describe('CustomizeList component', () => {
  const state = {
    features: {
      Processor: [
        {
          name: '17th Generation Intel Core HB (7 Core with donut spare)',
          cost: 700,
        },
        {
          name: 'Professor X AMD Fire Breather with sidewinder technology',
          cost: 1200,
        },
      ],
      'Operating System': [
        {
          name: 'Ubuntu Linux 16.04',
          cost: 200,
        },
        {
          name: 'Bodhi Linux',
          cost: 300,
        },
      ],
      'Video Card': [
        {
          name: 'Toyota Corolla 1.5v',
          cost: 1150.98,
        },
        {
          name: 'Mind mild breeze 2000',
          cost: 1345,
        },
      ],
      Display: [
        {
          name: '15.6" UHD (3840 x 2160) 60Hz Bright Lights and Knobs',
          cost: 1500,
        },
        {
          name: '15.3" HGTV (3840 x 2160) Home makeover edition',
          cost: 1400,
        },
      ],
    },
    selected: {
      Processor: {
        name: '17th Generation Intel Core HB (7 Core with donut spare)',
        cost: 700,
      },
      'Operating System': {
        name: 'Ubuntu Linux 16.04',
        cost: 200,
      },
      'Video Card': {
        name: 'Toyota Corolla 1.5v',
        cost: 1150.98,
      },
      Display: {
        name: '15.6" UHD (3840 x 2160) 60Hz Bright Lights and Knobs',
        cost: 1500,
      },
    },
  }
  it('should render without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <CustomizeList
        features={state.features}
        formatUS={USCurrencyFormat}
        selected={state.selected}
        updateFeature={(feature, newValue) =>
          updateFeature(feature, newValue)
        }
      />,
      div
    )
    ReactDOM.unmountComponentAtNode(div)
  })

  it('renders same UI', () => {
    const wrapper = shallow(
      <CustomizeList
      features={state.features}
      formatUS={USCurrencyFormat}
      selected={state.selected}
      updateFeature={(feature, newValue) =>
        updateFeature(feature, newValue)
      }
      />
    )
    expect(toJSON(wrapper)).toMatchSnapshot()
  })
})
