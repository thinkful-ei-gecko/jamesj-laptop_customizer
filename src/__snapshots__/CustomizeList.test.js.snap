// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomizeList component renders same UI 1`] = `
Array [
  <fieldset
    className="feature"
    key="Processor-0"
  >
    <legend
      className="feature__name"
    >
      <h3>
        Processor
      </h3>
    </legend>
    <CustomizeItem
      cost={700}
      feature="Processor"
      formatUS={NumberFormat {}}
      item={
        Object {
          "cost": 700,
          "name": "17th Generation Intel Core HB (7 Core with donut spare)",
        }
      }
      itemHash="\\"name\\":\\"17th-Generation-Intel-Core-HB-(7-Core-with-donut-spare)\\"\\"cost\\":700"
      key="\\"name\\":\\"17th-Generation-Intel-Core-HB-(7-Core-with-donut-spare)\\"\\"cost\\":700"
      name="17th Generation Intel Core HB (7 Core with donut spare)"
      selected={
        Object {
          "Display": Object {
            "cost": 1500,
            "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
          },
          "Operating System": Object {
            "cost": 200,
            "name": "Ubuntu Linux 16.04",
          },
          "Processor": Object {
            "cost": 700,
            "name": "17th Generation Intel Core HB (7 Core with donut spare)",
          },
          "Video Card": Object {
            "cost": 1150.98,
            "name": "Toyota Corolla 1.5v",
          },
        }
      }
      updateFeature={[Function]}
    />
    <CustomizeItem
      cost={1200}
      feature="Processor"
      formatUS={NumberFormat {}}
      item={
        Object {
          "cost": 1200,
          "name": "Professor X AMD Fire Breather with sidewinder technology",
        }
      }
      itemHash="\\"name\\":\\"Professor-X-AMD-Fire-Breather-with-sidewinder-technology\\"\\"cost\\":1200"
      key="\\"name\\":\\"Professor-X-AMD-Fire-Breather-with-sidewinder-technology\\"\\"cost\\":1200"
      name="Professor X AMD Fire Breather with sidewinder technology"
      selected={
        Object {
          "Display": Object {
            "cost": 1500,
            "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
          },
          "Operating System": Object {
            "cost": 200,
            "name": "Ubuntu Linux 16.04",
          },
          "Processor": Object {
            "cost": 700,
            "name": "17th Generation Intel Core HB (7 Core with donut spare)",
          },
          "Video Card": Object {
            "cost": 1150.98,
            "name": "Toyota Corolla 1.5v",
          },
        }
      }
      updateFeature={[Function]}
    />
  </fieldset>,
  <fieldset
    className="feature"
    key="Operating System-1"
  >
    <legend
      className="feature__name"
    >
      <h3>
        Operating System
      </h3>
    </legend>
    <CustomizeItem
      cost={200}
      feature="Operating System"
      formatUS={NumberFormat {}}
      item={
        Object {
          "cost": 200,
          "name": "Ubuntu Linux 16.04",
        }
      }
      itemHash="\\"name\\":\\"Ubuntu-Linux-16.04\\"\\"cost\\":200"
      key="\\"name\\":\\"Ubuntu-Linux-16.04\\"\\"cost\\":200"
      name="Ubuntu Linux 16.04"
      selected={
        Object {
          "Display": Object {
            "cost": 1500,
            "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
          },
          "Operating System": Object {
            "cost": 200,
            "name": "Ubuntu Linux 16.04",
          },
          "Processor": Object {
            "cost": 700,
            "name": "17th Generation Intel Core HB (7 Core with donut spare)",
          },
          "Video Card": Object {
            "cost": 1150.98,
            "name": "Toyota Corolla 1.5v",
          },
        }
      }
      updateFeature={[Function]}
    />
    <CustomizeItem
      cost={300}
      feature="Operating System"
      formatUS={NumberFormat {}}
      item={
        Object {
          "cost": 300,
          "name": "Bodhi Linux",
        }
      }
      itemHash="\\"name\\":\\"Bodhi-Linux\\"\\"cost\\":300"
      key="\\"name\\":\\"Bodhi-Linux\\"\\"cost\\":300"
      name="Bodhi Linux"
      selected={
        Object {
          "Display": Object {
            "cost": 1500,
            "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
          },
          "Operating System": Object {
            "cost": 200,
            "name": "Ubuntu Linux 16.04",
          },
          "Processor": Object {
            "cost": 700,
            "name": "17th Generation Intel Core HB (7 Core with donut spare)",
          },
          "Video Card": Object {
            "cost": 1150.98,
            "name": "Toyota Corolla 1.5v",
          },
        }
      }
      updateFeature={[Function]}
    />
  </fieldset>,
  <fieldset
    className="feature"
    key="Video Card-2"
  >
    <legend
      className="feature__name"
    >
      <h3>
        Video Card
      </h3>
    </legend>
    <CustomizeItem
      cost={1150.98}
      feature="Video Card"
      formatUS={NumberFormat {}}
      item={
        Object {
          "cost": 1150.98,
          "name": "Toyota Corolla 1.5v",
        }
      }
      itemHash="\\"name\\":\\"Toyota-Corolla-1.5v\\"\\"cost\\":1150.98"
      key="\\"name\\":\\"Toyota-Corolla-1.5v\\"\\"cost\\":1150.98"
      name="Toyota Corolla 1.5v"
      selected={
        Object {
          "Display": Object {
            "cost": 1500,
            "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
          },
          "Operating System": Object {
            "cost": 200,
            "name": "Ubuntu Linux 16.04",
          },
          "Processor": Object {
            "cost": 700,
            "name": "17th Generation Intel Core HB (7 Core with donut spare)",
          },
          "Video Card": Object {
            "cost": 1150.98,
            "name": "Toyota Corolla 1.5v",
          },
        }
      }
      updateFeature={[Function]}
    />
    <CustomizeItem
      cost={1345}
      feature="Video Card"
      formatUS={NumberFormat {}}
      item={
        Object {
          "cost": 1345,
          "name": "Mind mild breeze 2000",
        }
      }
      itemHash="\\"name\\":\\"Mind-mild-breeze-2000\\"\\"cost\\":1345"
      key="\\"name\\":\\"Mind-mild-breeze-2000\\"\\"cost\\":1345"
      name="Mind mild breeze 2000"
      selected={
        Object {
          "Display": Object {
            "cost": 1500,
            "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
          },
          "Operating System": Object {
            "cost": 200,
            "name": "Ubuntu Linux 16.04",
          },
          "Processor": Object {
            "cost": 700,
            "name": "17th Generation Intel Core HB (7 Core with donut spare)",
          },
          "Video Card": Object {
            "cost": 1150.98,
            "name": "Toyota Corolla 1.5v",
          },
        }
      }
      updateFeature={[Function]}
    />
  </fieldset>,
  <fieldset
    className="feature"
    key="Display-3"
  >
    <legend
      className="feature__name"
    >
      <h3>
        Display
      </h3>
    </legend>
    <CustomizeItem
      cost={1500}
      feature="Display"
      formatUS={NumberFormat {}}
      item={
        Object {
          "cost": 1500,
          "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
        }
      }
      itemHash="\\"name\\":\\"15.6\\"-UHD-(3840-x-2160)-60Hz-Bright-Lights-and-Knobs\\"\\"cost\\":1500"
      key="\\"name\\":\\"15.6\\"-UHD-(3840-x-2160)-60Hz-Bright-Lights-and-Knobs\\"\\"cost\\":1500"
      name="15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs"
      selected={
        Object {
          "Display": Object {
            "cost": 1500,
            "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
          },
          "Operating System": Object {
            "cost": 200,
            "name": "Ubuntu Linux 16.04",
          },
          "Processor": Object {
            "cost": 700,
            "name": "17th Generation Intel Core HB (7 Core with donut spare)",
          },
          "Video Card": Object {
            "cost": 1150.98,
            "name": "Toyota Corolla 1.5v",
          },
        }
      }
      updateFeature={[Function]}
    />
    <CustomizeItem
      cost={1400}
      feature="Display"
      formatUS={NumberFormat {}}
      item={
        Object {
          "cost": 1400,
          "name": "15.3\\" HGTV (3840 x 2160) Home makeover edition",
        }
      }
      itemHash="\\"name\\":\\"15.3\\"-HGTV-(3840-x-2160)-Home-makeover-edition\\"\\"cost\\":1400"
      key="\\"name\\":\\"15.3\\"-HGTV-(3840-x-2160)-Home-makeover-edition\\"\\"cost\\":1400"
      name="15.3\\" HGTV (3840 x 2160) Home makeover edition"
      selected={
        Object {
          "Display": Object {
            "cost": 1500,
            "name": "15.6\\" UHD (3840 x 2160) 60Hz Bright Lights and Knobs",
          },
          "Operating System": Object {
            "cost": 200,
            "name": "Ubuntu Linux 16.04",
          },
          "Processor": Object {
            "cost": 700,
            "name": "17th Generation Intel Core HB (7 Core with donut spare)",
          },
          "Video Card": Object {
            "cost": 1150.98,
            "name": "Toyota Corolla 1.5v",
          },
        }
      }
      updateFeature={[Function]}
    />
  </fieldset>,
]
`;
